src/main/java/sg/ncl/MainController.java

```diff
@@ -3341,6 +3347,162 @@ public String adminEnergy(Model model,
+            // ["A","B"] -> A,B
+            msg = msg.replace("[", "").replace("]", "").replace("\"", "");
+        }
+
+        redirectAttributes.addFlashAttribute(STATUS, status);
+        redirectAttributes.addFlashAttribute(MESSAGE, msg);
+    }
+
+    private void reserveNodes(ReservationStatusForm reservationStatusForm, RedirectAttributes redirectAttributes) {
+        HttpEntity<String> request = createHttpEntityHeaderOnly();
+        ResponseEntity response = null;
+
+        if (reservationStatusForm.getNumNodes() == null) {
+            redirectAttributes.addFlashAttribute(STATUS, NODES_RESERVATION_FAIL);
+            redirectAttributes.addFlashAttribute(MESSAGE, "Number of nodes not specify");
+            return;
+        }
+
+        if (reservationStatusForm.getMachineType() == null) {
+            // machine type not filled
+            // endpoint is the same as releaseNodes, i.e. /{id}
+            response = restTemplate.exchange(properties.reserveNodes(reservationStatusForm.getTeamId(), reservationStatusForm.getNumNodes(), null), HttpMethod.POST, request, String.class);
+        } else {
+            // reserve a specific number of machine type
+            response = restTemplate.exchange((properties.reserveNodes(reservationStatusForm.getTeamId(), reservationStatusForm.getNumNodes(), reservationStatusForm.getMachineType())), HttpMethod.POST, request, String.class);
+        }
+
+        JSONObject object = new JSONObject(response.getBody().toString());
+        String status = object.getString(STATUS);
+        String msg = object.optString(MESSAGE);
+
+        if ("nodes reservation OK".equals(status) && (msg.equals("[]"))) {
+            msg = "No nodes to be reserved";
+        } else if ("nodes reservation OK".equals(status)) {
+            // node ids
+            // ["A","B"] -> A,B
+            msg = msg.replace("[", "").replace("]", "").replace("\"", "");
+        }
+
+        redirectAttributes.addFlashAttribute(STATUS, status);
+        redirectAttributes.addFlashAttribute(MESSAGE, msg);
+    }
+
+    private List<String> convertJSONArrayToList(JSONArray jsonArray) {
+        List<String> resultList = new ArrayList<String>();
+        for (int i = 0; i < jsonArray.length(); i++) {
+            resultList.add(jsonArray.getString(i));
+        }
+        return resultList;
+    }
+
+    private HashMap<String, Team2> getTeamMap() {
+        TeamManager2 teamManager2 = new TeamManager2();
+        HttpEntity<String> request = createHttpEntityHeaderOnly();
+        ResponseEntity responseEntity = restTemplate.exchange(properties.getSioTeamsUrl(), HttpMethod.GET, request, String.class);
+
+        JSONArray jsonArray = new JSONArray(responseEntity.getBody().toString());
+
+        for (int i = 0; i < jsonArray.length(); i++) {
+            JSONObject jsonObject = jsonArray.getJSONObject(i);
+            Team2 one = extractTeamInfo(jsonObject.toString());
+            teamManager2.addTeamToTeamMap(one);
+        }
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace the type specification in this constructor call with the diamond operator ("<>"). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2293)